### PR TITLE
[HK] [Server] Hide non-randomized locations and exclude from hint calculations

### PR DESCRIPTION
**Pull request by Alchav**
Adds 'hidden' property to locations

Adds flag to item if it is set

'Hidden' locations / their items are not displayed in !missing, !checked or !remaining unless 'all' is specified as a parameter. They are excluded from hint point and hint point cost calculations.

Resolves Archipelago.HollowKnight issue 56
https://github.com/Ijwu/Archipelago.HollowKnight/issues/56

### CODE REVIEW

File: BaseClasses.py [hunk 1]
```diff
@@ -1041,6 +1041,7 @@ class Location:
     shop_slot_disabled: bool = False
     event: bool = False
     locked: bool = False
+    hidden = False
```
Comment (beauxq) [inline BaseClasses.py:1044]: To someone reading the code, it's not clear what "hidden" means.
"Does it mean this location is difficult to find? Does it mean the item at this location is invisible?"

You can put documentation on an attribute by putting a triple-quote string right below it.
```python3
    locked: bool = False
    hidden = False
    """ `hidden` means ... """
    game: str = "Generic"
```

Also, do we want type annotations on everything here? (My personal style would be that no annotation is needed if there's a clear type in the default value, but it looks like someone would like types on everything here.)

File: MultiServer.py [hunk 2]
```diff
@@ -503,7 +503,7 @@ def set_save(self, savedata: dict):
 
     def get_hint_cost(self, slot):
         if self.hint_cost:
-            return max(0, int(self.hint_cost * 0.01 * len(self.locations[slot])))
+            return max(0, int(self.hint_cost * 0.01 * len([loc for loc in self.locations[slot].values() if loc[2] & 8 == 0])))
```
Comment (Berserker66) [inline MultiServer.py:506]: this makes this normally very fast check (as len is cached) into a much slower version and it gets called a fair bit. A more efficient implementation might be good.

File: MultiServer.py [hunk 3]
```diff
@@ -792,10 +792,12 @@ def collect_player(ctx: Context, team: int, slot: int, is_group: bool = False):
                     collect_player(ctx, team, group, True)
 
 
-def get_remaining(ctx: Context, team: int, slot: int) -> typing.List[int]:
+def get_remaining(ctx: Context, team: int, slot: int, remove_hidden=False) -> typing.List[int]:
     items = []
     for location_id in ctx.locations[slot]:
         if location_id not in ctx.location_checks[team, slot]:
+            if remove_hidden and ctx.locations[slot][location_id][2] & 8 == 8:
```
Comment (Berserker66) [inline MultiServer.py:799]: & 8 should drop all non-8 bits, so the value is 0 or 8; meaning you should be able to drop the == 8

**Comment by Berserker66** [comment]
Left some code-relevant reviews, but I feel this potential change should be brought up with the community as well, do people want hidden locations in the first place?

**Comment by Alchav** [comment]
> Left some code-relevant reviews, but I feel this potential change should be brought up with the community as well, do people want hidden locations in the first place?

I suspect that people playing HK and not shuffling, for example, geo rocks would appreciate not having the over 200 geo rock locations flooding their !missing list, and needing to collect them anyways just for hint points.

Getting these items handled locally by the game would be a better solution, but that's not something I could figure out, and from the ap.hk issue page being tagged as "server-side" I assume that's not in the plans. But I have yet to hear from @Ijwu on this.

**Comment by Berserker66** [comment]
The way that I understood that is that the HK client will in the future read and use the already provided Randomize*poolname* options to make vanilla placements itself and in that case we drop it from the AP layer entirely when not randomized.

**Comment by JarnoWesthof** [comment]
Plz also add the new flag to the docs: https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/network%20protocol.md#NetworkItem

**Comment by Alchav** [comment]
> The way that I understood that is that the HK client will in the future read and use the already provided Randomize_poolname_ options to make vanilla placements itself and in that case we drop it from the AP layer entirely when not randomized.

We could always rip this code back out when that gets implemented

**Comment by black-sliver** [comment]
Not sure "hidden" is a good name, but if something is "unrandomized" because of player settings, than such a flag may be good. This could be useful for energy core in SoE for example, we can then always convert it to a chest and just pre-fill it and add the flag so it does not show up as a location.

**Comment by Berserker66** [comment]
> Not sure "hidden" is a good name, but if something is "unrandomized" because of player settings, then such a flag may be good. This could be useful for energy core in SoE for example, we can then always convert it to a chest and just pre-fill it and add the flag so it does not show up as a location.

which could be done currently by just None'ing out the location and item IDs, right?

My point I guess is, this can already be done with less processing/data accumulating, aka less feature bloat.

**Comment by black-sliver** [comment]
For evermore it would have to handle the special placement in two places (setting it to None, and then override during rom generation because we can't place None in the rom). With hidden it could generate the rom as normal, but tell the MultiServer to handle it as if it was a None. For full remote games this becomes a bigger difference.

**Comment by Alchav** [comment]
> Plz also add the new flag to the docs: https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/network%20protocol.md#NetworkItem

If the plan is to eventually make non-randomized HK locations local, then I would say we should plan on removing this code after that happens. In that case, I'd rather not encourage anyone to make clients to do anything with this flag, and instead add comments to my code indicating the plan to remove them in the future. Using the item flag bit is just an easier way to get this stored vs adding a new variable

**Pull request closed without merging**